Enumeration constants not allowed as case labels

```c
enum E { FOO = 42 };

int foo(enum E e) {
    switch (e) {
        case FOO:
            return 1;
    }
    return 0;
}
```

This should compile fine but it produces this error:

```
./test.c:5:14: error: case value must be an integer constant expression
        case FOO:
             ^
```

--- a/src/Parser.zig
+++ b/src/Parser.zig
@@ -1994,9 +1994,12 @@ const Enumerator = struct {
     res: Result,
 
     fn init(p: *Parser) Enumerator {
+        // Enumerator value is captured after increment in p.enumerator(), and we want the first enumeration constant
+        // to have a value of 0
+        const initial_value = Value.int(@as(i64, -1));
         return .{ .res = .{
             .ty = .{ .specifier = if (p.pp.comp.langopts.short_enums) .schar else .int },
-            .val = Value.int(0),
+            .val = initial_value,
         } };
     }
 
@@ -2046,7 +2049,7 @@ fn enumerator(p: *Parser, e: *Enumerator) Error!?EnumFieldAndNode {
     var res = e.res;
     res.ty = try p.withAttributes(res.ty, attr_buf_top);
 
-    try p.syms.defineEnumeration(p, res.ty, name_tok);
+    try p.syms.defineEnumeration(p, res.ty, name_tok, e.res.val);
     const node = try p.addNode(.{
         .tag = .enum_field_decl,
         .ty = res.ty,
@@ -5530,7 +5533,7 @@ fn primaryExpr(p: *Parser) Error!Result {
                     }
                 }
                 return Result{
-                    .val = if (p.const_decl_folding == .no_const_decl_folding) Value{} else sym.val,
+                    .val = if (p.const_decl_folding == .no_const_decl_folding and sym.kind != .enumeration) Value{} else sym.val,
                     .ty = sym.ty,
                     .node = try p.addNode(.{
                         .tag = if (sym.kind == .enumeration) .enumeration_ref else .decl_ref_expr,

--- a/src/SymbolStack.zig
+++ b/src/SymbolStack.zig
@@ -331,6 +331,7 @@ pub fn defineEnumeration(
     p: *Parser,
     ty: Type,
     tok: TokenIndex,
+    val: Value,
 ) !void {
     const name = p.tokSlice(tok);
     const kinds = s.syms.items(.kind);
@@ -358,6 +359,6 @@ pub fn defineEnumeration(
         .name = name,
         .tok = tok,
         .ty = ty,
-        .val = .{},
+        .val = val,
     });
 }

--- a/test/cases/containers.c
+++ b/test/cases/containers.c
@@ -128,8 +128,7 @@ struct A {
 struct B b1;
 
 
-#define EXPECTED_ERRORS "containers.c:11:9: error: enum value must be an integer constant expression" \
-    "containers.c:15:8: error: use of 'Foo' with tag type that does not match previous definition" \
+#define EXPECTED_ERRORS "containers.c:15:8: error: use of 'Foo' with tag type that does not match previous definition" \
     "containers.c:9:6: note: previous definition is here" \
     "containers.c:15:12: error: variable has incomplete type 'struct Foo'" \
     "containers.c:20:6: warning: declaration does not declare anything [-Wmissing-declaration]" \

--- a/test/cases/enumerator constants.c
+++ b/test/cases/enumerator constants.c
@@ -1,0 +1,22 @@
+enum E {
+	A,
+	B = 10,
+	C,
+	D = 2,
+	E = -2,
+	F,
+};
+
+_Static_assert(A == 0, "A is wrong");
+_Static_assert(B == 10, "B is wrong");
+_Static_assert(C == 11, "C is wrong");
+_Static_assert(D == 2, "D is wrong");
+_Static_assert(E == -2, "E is wrong");
+_Static_assert(F == -1, "F is wrong");
+
+void foo(enum E e) {
+	switch (e) {
+		case A: return;
+		default: return;
+	}
+}

